RFC: Time Series Culling

mrtracy opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9343)

<!-- Reviewable:end -->

Commit:
RFC: Time Series Culling

bdarnell commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/time_series_culling.md, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbIhY1tVZx8ReM93O8:-KRbIhY2fLiTYs_04Zrs:bqwd0pl) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L1)):*
> ```Markdown
> - Feature Name: Time Series Rollups
> ```

The filename and feature name should match.

---

*[docs/RFCS/time_series_culling.md, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbJR85T0AvWFSxbghZ:-KRbJR85T0AvWFSxbgh_:bsm19q) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L36)):*
> ```Markdown
> # Motivation
> 
> In our test clusters, this creates a very large amount of data (on the order of
> ```

How much does an idle single node cluster produce? Is this data volume just a function of node count and time or do other factors contribute?

---

*[docs/RFCS/time_series_culling.md, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbKJVlsCgQmsbk5waj:-KRbKJVmkicaB1-0bRXW:bmp1kgw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L49)):*
> ```Markdown
> 
> To give some context of numbers: currently, all metrics on disk are stored in a
> format which is downsampled to _ten second sample periods_; this is the
> ```

I'll repeat my claim that 10-second intervals are unnecessary and we should consider increasing the interval to 30s or 60s. For comparison, prometheus has a [default](https://prometheus.io/docs/operating/configuration/#configuration-file) of 60s, although many of their [examples](https://prometheus.io/docs/introduction/getting_started/#configuring-prometheus-to-monitor-itself) use 15s. 

Of course, changing the rate of growth doesn't change the fact that we need to discard old data to prevent unbounded growth. 

---

*[docs/RFCS/time_series_culling.md, line 103 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbNtcQmAAO7uQ4wpLi:-KRbNtcRwiuDNmlmingf:b-7ol1id) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L103)):*
> ```Markdown
> x 24 x 14) = *3360* keys that should not be culled. In a system that maintains
> dozens, possibly hundreds of time series, this is a lot of data for each node to
> scan on a regular basis.
> ```

Is it? It would be a lot for every node to scan the entire TS range, but if we had a queue like the `gcQueue` or the `replicaConsistencyQueue` so that each node periodically processed the TS ranges for which it is the leader, it seems reasonable. 

---

*[docs/RFCS/time_series_culling.md, line 122 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbOR91YSjxKdNSeHsF:-KRbOR91YSjxKdNSeHsG:bazp30t) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L122)):*
> ```Markdown
> that any are removed. This requires some manual effort on the part of
> developers; however, the consequences for failing to do so are not especially
> severe (a limited amount of old data will persist on the cluster).
> ```

Alternately, we could persist the set of used timeseries names, which avoids the manual effort at the expense of some complexity to save and restore the list. 

---

*[docs/RFCS/time_series_culling.md, line 155 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbOlOwgJH9q4MEm9PS:-KRbOlOwgJH9q4MEm9PT:bedn24k) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L155)):*
> ```Markdown
> #### Culling low resolution data
> 
> Although the volume is exceptionally lower, low-resolution data will still build
> ```

s/exponentially/much/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/time_series_culling.md, line 61 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRcvteD-cBGQ9i8W4t6:-KRcvteESXMaOqmVKUQO:b-qapj54) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L61)):*
> ```Markdown
> Thus, the expected data storage of the low-resolution is approximately _180x
> smaller_ than the high-resolution (not 360 because the individual low-resolution
> samples will include a "min" and "max" value not present at the high-resolution)
> ```

Are sum and count values present at high resolution? They are described as part of the rollup algorithm.

---

*[docs/RFCS/time_series_culling.md, line 122 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbOR91YSjxKdNSeHsF:-KRcsjtDo-MG1bLgoFKt:b7i1yor) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L122)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Alternately, we could persist the set of used timeseries names, which avoids the manual effort at the expense of some complexity to save and restore the list. 
</details>
Is it possible to efficiently recover the list of time series? 

1) Do a scan of length 1 starting at `[ts prefix]`.
2) Extract the series name out of the result.
3) Compute the next key of `[ts prefix][series name]`.
4) Do a scan of length 1 at that next key. Goto 2 until `[ts prefix]` is exceeded.

This would avoid any of the list bookkeeping. 

---

*[docs/RFCS/time_series_culling.md, line 155 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRcvBg2NHx7UKtsNEP_:-KRcvBg30X2VttBHOK8t:b-gfui78) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L155)):*
> ```Markdown
> #### Culling low resolution data
> 
> Although the volume is exceptionally lower, low-resolution data will still build
> ```

It says exceptionally, not exponentially.

---

*[docs/RFCS/time_series_culling.md, line 172 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRcwUijTPr10Tl3zwr4:-KRcwUijTPr10Tl3zwr5:b9i2l0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L172)):*
> ```Markdown
> 3. Write the computed sample as a low-resolution data point into the time series
> system; this is exactly the same process as currently recorded time series,
> except it will be writing to a different key space (with a different key
> ```

Do we need to describe this new key space here or is that not useful because it's already obvious how to do that (i.e., to those already familiar with the TS system)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/time_series_culling.md, line 86 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRd0hz0kzb2_KLZk2hY:-KRd0hz0kzb2_KLZk2hZ:b-sut0wh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L86)):*
> ```Markdown
> Locating time series keys to be culled is not completely trivial due to the
> construction of time series keys, which is thus: 
> `[ts prefix][series name][timestamp][source]`
> ```

Having not read the entire doc, this comment might be inappropriate: I was imagining time series culling would be performed by a replica scanner queue. The "time series GC" queue would require the range lease be held (and thus only a single node in the cluster will perform the culling for a given range) and peek at the start and end keys to see if the range contained any time series data. Similar to the consistency checker we could perform time series GC at a cadence of once per day. 

When processing a range it would scan over all of the keys and build up a batch containing `DeleteRange` requests for any keys that are too old. 

With this approach there is no need to maintain a list of time series keys. The "time series GC" queue is guaranteed to see all ranges. Eventually this could mesh nicely with roll-up. In addition to invoking the `DeleteRange` the queue would write the rolled-up data. 

---

*[docs/RFCS/time_series_culling.md, line 103 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbNtcQmAAO7uQ4wpLi:-KRd1iDYqz250ji4Tgu1:bx8m89) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L103)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Is it? It would be a lot for every node to scan the entire TS range, but if we had a queue like the `gcQueue` or the `replicaConsistencyQueue` so that each node periodically processed the TS ranges for which it is the leader, it seems reasonable. 
</details>
Ah, I see Ben has thought of the queue approach as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/time_series_culling.md, line 122 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbOR91YSjxKdNSeHsF:-KRdKUKVc45BDqNw4PjB:b-dodl75) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L122)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> Is it possible to efficiently recover the list of time series? 
> 
> 1) Do a scan of length 1 starting at `[ts prefix]`.
> 2) Extract the series name out of the result.
> 3) Compute the next key of `[ts prefix][series name]`.
> 4) Do a scan of length 1 at that next key. Goto 2 until `[ts prefix]` is exceeded.
> 
> This would avoid any of the list bookkeeping. 
</details>
It's possible to do that, but I don't know if I'd call it efficient. All those serialized RPCs are not obviously better than just scanning all the data and throwing most of it away (especially if we can use some server-side smarts like we'd use for distributed sql). 

---

*[docs/RFCS/time_series_culling.md, line 155 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRcvBg2NHx7UKtsNEP_:-KRdKjdMgRj-UJ0grIhI:b-wfn10b) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L155)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> It says exceptionally, not exponentially.
</details>
Ah, so it does. s/exceptionally/much/, then :) (what's the rule that this is an exception to?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/time_series_culling.md, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbKJVlsCgQmsbk5waj:-KRdU-hQ0kB96nSG4jJM:brokepk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L49)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I'll repeat my claim that 10-second intervals are unnecessary and we should consider increasing the interval to 30s or 60s. For comparison, prometheus has a [default](https://prometheus.io/docs/operating/configuration/#configuration-file) of 60s, although many of their [examples](https://prometheus.io/docs/introduction/getting_started/#configuring-prometheus-to-monitor-itself) use 15s. 
> 
> Of course, changing the rate of growth doesn't change the fact that we need to discard old data to prevent unbounded growth. 
</details>
Yeah, I've never been in a situation where I found that the available 30s or 60s granularity metrics weren't enough. Once you narrow in far enough on when a problem started, it's typically time to start digging through logs in my experience.

---

*[docs/RFCS/time_series_culling.md, line 54 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRdUdfcYYNbGsCl-0Oz:-KRdUdfcYYNbGsCl-0P-:bkb4jly) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L54)):*
> ```Markdown
> We also propose that, when this data is deleted, it is first downsampled further
> into _one hour sample periods_; this is the "low-resolution" data. This data
> will be kept for a much longer time, likely _6-12 months_, but perhaps longer.
> ```

It'd be nice to see a back of the envelope calculation of how much data this will be in a few different cluster sizes.

---

*[docs/RFCS/time_series_culling.md, line 128 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRdW8R5MPwbrr9If6be:-KRdW8R5MPwbrr9If6bf:b-qhjfyg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L128)):*
> ```Markdown
> Distribution of the culling process across the cluster is another important
> concern. We want to *guarantee* that all time series are eventually culled. We
> also want time series to be culled quickly, which can be helped by properly
> ```

Why do we want them to be culled quickly? Is it really a big deal if the data sticks around an extra couple hours?

---

*[docs/RFCS/time_series_culling.md, line 196 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRdXC8HJEVlsiQ2b-bn:-KRdXC8HJEVlsiQ2b-bo:b-ndtubc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L196)):*
> ```Markdown
> of modifying low-level iterators and querying slightly different data. This
> component will likely be the most complicated to actually *write*, but it should
> be somewhat easier to *test* that the above algorithms, as there is already
> ```

s/that/than

---

*[docs/RFCS/time_series_culling.md, line 206 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRdXgtOi1PHqnAKAcib:-KRdXgtOi1PHqnAKAcic:b7ih4fq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L206)):*
> ```Markdown
> # Alternatives
> 
> ### Delete Only
> ```

I suppose it's fairly obvious in this case, but it's typically worth mentioning why the alternative wasn't chosen

---

*[docs/RFCS/time_series_culling.md, line 236 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRdXZhvG45JMLzGe12U:-KRdXZhvG45JMLzGe12V:bgtrqnc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L236)):*
> ```Markdown
> 
> Instead of rolling up when low-resolution data is deleted, it is instead rolled
> up as soon as entire hour of high-resolution samples has been collected in a
> ```

s/as entire/as an entire/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 0 of 1 files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/time_series_culling.md, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbIhY1tVZx8ReM93O8:-KRel8Kdht_da1DDRnp-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L1)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The filename and feature name should match.
</details>
Done.

---

*[docs/RFCS/time_series_culling.md, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbJR85T0AvWFSxbghZ:-KRelgiPBNLwbhRyOh-b:b2k3mks) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L36)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> How much does an idle single node cluster produce? Is this data volume just a function of node count and time or do other factors contribute?
</details>
Added a note that it is indeed a factor of the node count.

---

*[docs/RFCS/time_series_culling.md, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbKJVlsCgQmsbk5waj:-KRellO7NCB_gEt7n8xE:b-4wcqsp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L49)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> Yeah, I've never been in a situation where I found that the available 30s or 60s granularity metrics weren't enough. Once you narrow in far enough on when a problem started, it's typically time to start digging through logs in my experience.
</details>
With the culling system in place, this rate can be modified down the road. However, it does seem to be a judgement call, and there are going to be varying opinions in the absence of hard data on the exact load of time series.

---

*[docs/RFCS/time_series_culling.md, line 61 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRcvteD-cBGQ9i8W4t6:-KRem_0MTAzN1u39MmEq:btgva92) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L61)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> Are sum and count values present at high resolution? They are described as part of the rollup algorithm.
</details>
Sum and count are present on all resolutions; min and max will only be present on rollup resolutions.

---

*[docs/RFCS/time_series_culling.md, line 86 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRd0hz0kzb2_KLZk2hY:-KRf-20HZzmM8BqZqxzq:brv8tlx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L86)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Having not read the entire doc, this comment might be inappropriate: I was imagining time series culling would be performed by a replica scanner queue. The "time series GC" queue would require the range lease be held (and thus only a single node in the cluster will perform the culling for a given range) and peek at the start and end keys to see if the range contained any time series data. Similar to the consistency checker we could perform time series GC at a cadence of once per day. 
> 
> When processing a range it would scan over all of the keys and build up a batch containing `DeleteRange` requests for any keys that are too old. 
> 
> With this approach there is no need to maintain a list of time series keys. The "time series GC" queue is guaranteed to see all ranges. Eventually this could mesh nicely with roll-up. In addition to invoking the `DeleteRange` the queue would write the rolled-up data. 
</details>
Based on feedback from you and Ben, I have changed this plan to use a queue-based system.  I have outlined the algorithm needed, but have also added a section about my concerns re: package dependency.

I have moved the original location algorithm to the alternatives section.

---

*[docs/RFCS/time_series_culling.md, line 103 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbNtcQmAAO7uQ4wpLi:-KRf-rRruKvrotv_f_kb:b-q64pkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L103)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Ah, I see Ben has thought of the queue approach as well.
</details>
See above.

---

*[docs/RFCS/time_series_culling.md, line 122 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbOR91YSjxKdNSeHsF:-KRf-sLD0Tp5u2VuZ1gu:b-q64pkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L122)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> It's possible to do that, but I don't know if I'd call it efficient. All those serialized RPCs are not obviously better than just scanning all the data and throwing most of it away (especially if we can use some server-side smarts like we'd use for distributed sql). 
</details>
See above.

---

*[docs/RFCS/time_series_culling.md, line 128 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRdW8R5MPwbrr9If6be:-KRf-tSd5sVEi_TPQYSk:b-piesw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L128)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> Why do we want them to be culled quickly? Is it really a big deal if the data sticks around an extra couple hours?
</details>
Section has been removed.

---

*[docs/RFCS/time_series_culling.md, line 155 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRcvBg2NHx7UKtsNEP_:-KRenxcL3a9FNTrAgbfh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L155)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Ah, so it does. s/exceptionally/much/, then :) (what's the rule that this is an exception to?)
</details>
Done.

---

*[docs/RFCS/time_series_culling.md, line 172 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRcwUijTPr10Tl3zwr4:-KRenWZI_pBTmd3oBsMp:bo4kq3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L172)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> Do we need to describe this new key space here or is that not useful because it's already obvious how to do that (i.e., to those already familiar with the TS system)?
</details>
The `[ts prefix]` mentioned above is actually a combination of something like "\0x04ts" and a resolution-specific three letter code, something like "10s". I have omitted this from the document, as you have guessed, because it is a documented feature of the time series system (there is quite a bit written in the doc.go in the ts package).

---

*[docs/RFCS/time_series_culling.md, line 196 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRdXC8HJEVlsiQ2b-bn:-KReomz_CP_6n2EwhlGx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L196)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> s/that/than
</details>
Done.

---

*[docs/RFCS/time_series_culling.md, line 206 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRdXgtOi1PHqnAKAcib:-KRezssW66kzgKIGup9N:br8o348) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L206)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> I suppose it's fairly obvious in this case, but it's typically worth mentioning why the alternative wasn't chosen
</details>
This is no longer an alternative, it will be a phase of implementation.  I have added a section on this phased implementation above.

---

*[docs/RFCS/time_series_culling.md, line 236 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRdXZhvG45JMLzGe12U:-KRezsEvx8VN-1-xq996:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L236)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> s/as entire/as an entire/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 0 of 1 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/time_series_culling.md, line 54 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRdUdfcYYNbGsCl-0Oz:-KRf0PEj5eGzVJhJim42:b7sq858) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L54)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> It'd be nice to see a back of the envelope calculation of how much data this will be in a few different cluster sizes.
</details>
I've added some notes above; the growth is basically a flat function of (#nodes x #time series).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/time_series_culling.md, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbJR85T0AvWFSxbghZ:-KRfQ259z1LS2bHvHqnQ:b6xrwd7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L36)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> Added a note that it is indeed a factor of the node count.
</details>
Can you track down the calculation (or re-perform it) regarding how much data a single metric produces per hour as well as the current number of metrics per store and per node? Together those would make the "several gigabytes per week" statement more concrete and it seems reasonable to have those values live here instead of in some random issue. Happy to try and track that info down if you can't find it (or reproduce it).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/time_series_culling.md, line 121 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRfQj_nn06xMMnX-Llz:-KRfQj_nn06xMMnX-Lm-:b-ht1h99) ([raw file](https://github.com/cockroachdb/cockroach/blob/953fcada0f9c9ff564ec4d9b071779a8be34a59d/docs/RFCS/time_series_culling.md#L121)):*
> ```Markdown
> follows:
> 
> 1. Find for the first time series key in the range (scan for [ts prefix]). 
> ```

`s/for the/the/g`

---

*[docs/RFCS/time_series_culling.md, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRfQo7BUKM0caEYwi8P:-KRfQo7BUKM0caEYwi8Q:bbi73pf) ([raw file](https://github.com/cockroachdb/cockroach/blob/953fcada0f9c9ff564ec4d9b071779a8be34a59d/docs/RFCS/time_series_culling.md#L123)):*
> ```Markdown
> 1. Find for the first time series key in the range (scan for [ts prefix]). 
> 2. Deconstruct the key to retrieve its name.
> 3. Run the rollup operation on all keys in the range: 
> ```

`s,rollup,rollup/delete,g`

---

*[docs/RFCS/time_series_culling.md, line 136 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRfR0NAHPryMvGzpoWk:-KRfR0NAHPryMvGzpoWl:bt2hgf5) ([raw file](https://github.com/cockroachdb/cockroach/blob/953fcada0f9c9ff564ec4d9b071779a8be34a59d/docs/RFCS/time_series_culling.md#L136)):*
> ```Markdown
> in order to avoid duplicate work; however, this is *not* necessary for
> correctness, as demonstrated in the [Rollup Algorithm](#rollup-algorithm)
> section below.
> ```

At what cadence will the queue examine ranges?

---

*[docs/RFCS/time_series_culling.md, line 214 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRfQa4TaBQMTbdpyTME:-KRfQa4TaBQMTbdpyTMF:ba4dxmn) ([raw file](https://github.com/cockroachdb/cockroach/blob/953fcada0f9c9ff564ec4d9b071779a8be34a59d/docs/RFCS/time_series_culling.md#L214)):*
> ```Markdown
> limiting the growth of time series data on the cluster.
> 
> 2. The "rollup" algorithm will be implementing, generating low-resolution data
> ```

`s/implementing/implemented/g`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-:-KRfRbcJAuIKt8DDElps:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 1 files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/time_series_culling.md, line 9 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRfREgnfmxmULgFB3Nk:-KRfREgnfmxmULgFB3Nl:bahwiuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/953fcada0f9c9ff564ec4d9b071779a8be34a59d/docs/RFCS/time_series_culling.md#L9)):*
> ```Markdown
> 
> # Summary
> Currently, Time Series data recorded for by CockroachDB for its own internal
> ```

"for by"

---

*[docs/RFCS/time_series_culling.md, line 12 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRfRK7ULfRvY2uZf2wi:-KRfRK7ULfRvY2uZf2wj:b-bia3zu) ([raw file](https://github.com/cockroachdb/cockroach/blob/953fcada0f9c9ff564ec4d9b071779a8be34a59d/docs/RFCS/time_series_culling.md#L12)):*
> ```Markdown
> metrics is retained indefinitely. High-resolution metrics data quickly loses 
> utility as it ages, consuming disk space and creating range-related overhead
> without confering an appropriate benefit.
> ```

conferring

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 1 files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/time_series_culling.md, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbKJVlsCgQmsbk5waj:-KRiDOo0x8Wp4G1mZGdN:b8r8oto) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L49)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> With the culling system in place, this rate can be modified down the road. However, it does seem to be a judgement call, and there are going to be varying opinions in the absence of hard data on the exact load of time series.
</details>
It will be valuable understanding the amount of data per metric here? Every 10 seconds is 3153600 samples a year per metric. 

---

*[docs/RFCS/time_series_culling.md, line 54 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRdUdfcYYNbGsCl-0Oz:-KRiEfYmRN55cSQzAAH6:b-tiljd4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L54)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> I've added some notes above; the growth is basically a flat function of (#nodes x #time series).
</details>
If you down sample the 3153600 samples at a 1 hour interval, thats  8760  samples per year per metric (per node).  Would it be acceptable to down sample to 5 minutes?

---

*[docs/RFCS/time_series_culling.md, line 164 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRiJFsXE3plta4WZFQ4:-KRiJFsYs_1ltEglaXVI:b-la4gyb) ([raw file](https://github.com/cockroachdb/cockroach/blob/953fcada0f9c9ff564ec4d9b071779a8be34a59d/docs/RFCS/time_series_culling.md#L164)):*
> ```Markdown
> Although the volume is much lower, low-resolution data will still build
> up indefinitely unless it is culled. This data could also be culled by the same
> algorithm outlined here; it could then be rolled-up further, or simply deleted.
> ```

deleted!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-:-KRiN7z9SUOh775SreiX:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 0 of 1 files reviewed at latest revision, 19 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/time_series_culling.md, line 9 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRfREgnfmxmULgFB3Nk:-KS-_0fOXEJPt7w025zd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/953fcada0f9c9ff564ec4d9b071779a8be34a59d/docs/RFCS/time_series_culling.md#L9)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "for by"
</details>
Done.

---

*[docs/RFCS/time_series_culling.md, line 12 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRfRK7ULfRvY2uZf2wi:-KS-_15PWSMx9E5jDlrK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/953fcada0f9c9ff564ec4d9b071779a8be34a59d/docs/RFCS/time_series_culling.md#L12)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> conferring
</details>
Done.

---

*[docs/RFCS/time_series_culling.md, line 121 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRfQj_nn06xMMnX-Llz:-KS-bGGkilRxCNqez3PS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/953fcada0f9c9ff564ec4d9b071779a8be34a59d/docs/RFCS/time_series_culling.md#L121)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/for the/the/g`
</details>
Done.

---

*[docs/RFCS/time_series_culling.md, line 123 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRfQo7BUKM0caEYwi8P:-KS-bY8glnfkNjYI3Td2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/953fcada0f9c9ff564ec4d9b071779a8be34a59d/docs/RFCS/time_series_culling.md#L123)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s,rollup,rollup/delete,g`
</details>
Done.

---

*[docs/RFCS/time_series_culling.md, line 136 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRfR0NAHPryMvGzpoWk:-KS-a2ceivh2_RYaznkk:b-jhl0bg) ([raw file](https://github.com/cockroachdb/cockroach/blob/953fcada0f9c9ff564ec4d9b071779a8be34a59d/docs/RFCS/time_series_culling.md#L136)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> At what cadence will the queue examine ranges?
</details>
Didn't realize this was set per range. We can use the same timing as the replica GC queue initially, that makes sense to me; we could set it lower if it becomes a problem.

---

*[docs/RFCS/time_series_culling.md, line 214 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRfQa4TaBQMTbdpyTME:-KS-aFes2OG7E4A-MjYP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/953fcada0f9c9ff564ec4d9b071779a8be34a59d/docs/RFCS/time_series_culling.md#L214)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/implementing/implemented/g`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 0 of 1 files reviewed at latest revision, 19 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/time_series_culling.md, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbJR85T0AvWFSxbghZ:-KScXTFX_Ct0IBG9uS-h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L36)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Can you track down the calculation (or re-perform it) regarding how much data a single metric produces per hour as well as the current number of metrics per store and per node? Together those would make the "several gigabytes per week" statement more concrete and it seems reasonable to have those values live here instead of in some random issue. Happy to try and track that info down if you can't find it (or reproduce it).
</details>
Done.

---

*[docs/RFCS/time_series_culling.md, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbKJVlsCgQmsbk5waj:-KScXVp2qaArNXDGAgj8:bkmsiv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L49)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> It will be valuable understanding the amount of data per metric here? Every 10 seconds is 3153600 samples a year per metric. 
</details>
Added information about the amount of data being generated above.

---

*[docs/RFCS/time_series_culling.md, line 164 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRiJFsXE3plta4WZFQ4:-KScXa33pWJ2jhVr3Y-a:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/953fcada0f9c9ff564ec4d9b071779a8be34a59d/docs/RFCS/time_series_culling.md#L164)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> deleted!
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/time_series_culling.md, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbKJVlsCgQmsbk5waj:-KSd5XBMt21x6MN0iMUC:bhb9cc2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L49)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> Added information about the amount of data being generated above.
</details>
Given the numbers you calculated, perhaps it would be a good time to reduce the sample rate to 60s. We'd still want culling, but a 6x reduction in timeseries data sounds useful. I also recall thinking that the low-level timeseries storage had some bloat. The number 20% comes to mind, but I don't recall where from (something to do with the proto encoding). I think we could switch that format if we're willing to drop all of the old data (we just switch to a different timeseries key prefix).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 0 of 1 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/time_series_culling.md, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbKJVlsCgQmsbk5waj:-KSdC39hDL14vV9ftIE6:b4f9ut6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L49)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Given the numbers you calculated, perhaps it would be a good time to reduce the sample rate to 60s. We'd still want culling, but a 6x reduction in timeseries data sounds useful. I also recall thinking that the low-level timeseries storage had some bloat. The number 20% comes to mind, but I don't recall where from (something to do with the proto encoding). I think we could switch that format if we're willing to drop all of the old data (we just switch to a different timeseries key prefix).
</details>
I've been in rare situations where there have been intermittent << 1 minute spikes of latency, caused by spikes in SQL queries (or the occasional slow query that monopolizes a database). In those cases, having frequent metrics samples helped diagnose the problem more quickly. If we had a slow query log, you could imagine correlating entries in the slow query log with increases in query latency to determine the cause  of the slowdown. So, I think 60 seconds is actually too infrequent for some cases.

I don't know how representative my experiences are, but I think the sampling interval should be configurable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 0 of 1 files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/time_series_culling.md, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KRbKJVlsCgQmsbk5waj:-KSgKlb6Kht_a9S1CxdA:b-goknao) ([raw file](https://github.com/cockroachdb/cockroach/blob/3159a21801cb0dfb35cff839a8e0771fd61e2eba/docs/RFCS/time_series_culling.md#L49)):*
<details><summary><i>Previously, cuongdo (Cuong Do) wrote…</i></summary>
> I've been in rare situations where there have been intermittent << 1 minute spikes of latency, caused by spikes in SQL queries (or the occasional slow query that monopolizes a database). In those cases, having frequent metrics samples helped diagnose the problem more quickly. If we had a slow query log, you could imagine correlating entries in the slow query log with increases in query latency to determine the cause  of the slowdown. So, I think 60 seconds is actually too infrequent for some cases.
> 
> I don't know how representative my experiences are, but I think the sampling interval should be configurable.
</details>
I had the opposite impression when I computed the numbers, actually; my impression was "Is even a terabyte of data too much for a product intended to hold petabytes?"  200MB, per node, per week is the current number; 600 MB if you count replicas. If we cull after a month, that means we are asking for 3GB of disk space per node for metrics data.  We can reduce this number by culling sooner (e.g. 2 week cull = 1.5GB per node) or reducing the resolution of the original sampling (60s over a month = 500MB per node), or both (2 weeks, 60s = 250MB per node).

This does not have an impact on this specific RFC, so I am going to commit it and begin work under the assumption that we are keeping data at 10s resolution.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on docs/RFCS/time_series_culling.md:
```diff
@@ -0,0 +1,349 @@
+    `[ts prefix][series name][0] - [tsprefix][series name][now - threshold]`
+4. Find the next key on the range which contains data for a different time
+series by searching for key `PrefixEnd([ts prefix][series name])`.
+5. If a key was found in step 5, return to step 2 with that series name.
```
Judging from the beginnings of this in #9767, is this what would happen in the queue's `process`?

- find all the time series names (not only the first one) from the engine
- Iterating over them, generating and sending out for each a batch which contains the Rollout and the DeleteRange?

I'm asking because the `Rollout` and the `DeleteRange` should be tied together as closely as possible, and that (at least potentially) forces `DistSender` to put them into a transaction (due to spanning ranges). This necessitates intents on the timeseries keys, which is a whole can of worms we should stay away from.

So instead of that, you are going to have to send two individual batches, presumably doing the rollup first, and then the deletion. I think that's ok (though see comments below), but wanted to point it out as that's a good one to waste time on during imlementation.

tschottdorf commented on docs/RFCS/time_series_culling.md:
```diff
@@ -0,0 +1,349 @@
+
+#### Culling low resolution data
+
+Although the volume is much lower, low-resolution data will still build
```
Those are essentially folded up into the exact same mechanism where the culling of an unculled series results in a single culled sample at the new resolution, and the culling of a culled one results in zero samples generated, right?

tschottdorf commented on docs/RFCS/time_series_culling.md:
```diff
@@ -0,0 +1,349 @@
+before deleting the high-resolution data. However, the low-resolution data will
+not immediately be accessible for queries.
+
+3. The query system will be modified to consider the high-resolution data.
```
s/high/low/

tschottdorf commented on docs/RFCS/time_series_culling.md:
```diff
@@ -0,0 +1,349 @@
+prefix).
+4. Delete the original high-resolution key.
+
+This algorithm is safe to use, even in the case where the same key is being
```
This isn't really correct, though I don't think we need to do anything but document that incorrectness and move on.

Consider that two replicas of the queue groom a time series concurrently. In reality, I think that those could even be two different replicas. For example, what if a cullable tail of a time series covers ten Ranges and each lease holder picks up the range at the same time? Many fun interleavings are possible. Let's limit to concurrency two:

- Replica 1 and Replica 2 get processed, discover that they have the time series.
- Replica 1 issues a Scan to get the keys to be rolled up, writes the roll-up, and deletes the keys.
- Replica 2 issues a Scan to get the keys to be rolled up. Wait, there aren't any! Let me write an empty sample and delete the keys (of which there were none).

Now you've created and immediately deleted the roll-up.

It should be very rare that two Replicas of the same Range race in this way. I'm less sure about time series which span ranges, or maybe I'm operating on false assumptions on how that works in practice.

In either case, there's something to discuss and document here.

tschottdorf commented on docs/RFCS/time_series_culling.md:
```diff
@@ -0,0 +1,349 @@
+culled by multiple nodes at the same time; this is because step 3 and 4 are
+currently *idempotent*. The low-resolution sample generated by each node will be
+identical, and the engine-level time series merging system currently discards
+duplicate samples. The deletion of the high-resolution key may cause an error on
```
Ah, this seems to deal with some of it. Are we requiring that the top-boundary of the key is there? Or are we simply requiring that the result set is not empty?

mrtracy commented:
Review status: 0 of 1 files reviewed at latest revision, 21 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/time_series_culling.md, line 160 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KR_IiHSXH02YhQDjcJZ-r3-160:-KTjfPvUkNO839XHod7i:b-5b3y36) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4acd5bc13a54a08259b1287a148d2db05e39887/docs/RFCS/time_series_culling.md#L160)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> And more concretely, after finding (say) the time series `ts/cr.node.sql.inserts/<cullable_ts>/(.*)` you send a scan for that range and make the rollup - what is the schema of the key that this goes to? A completely different keyspace? Would it be something like
> 
> 
> 
> `/ts_culled/cr.node.sql.inserts/1473739200/`?
> 
> 
> 
> I'm asking because it seems to make sense to put the culled data approximately where the old data was, i.e. write keys `ts/cr.node.sql.inserts/1473739200/(.*)` again, though that raises the question of how these are distinguished from unculled time series samples (maybe there's no need to?)
</details>
They are written close to each other.  Time series keys actually have the form:

`[ts prefix][series name][*resolution token*][timestamp][source]`

Meaning that high-resolution keys have the form `/ts/cr.node.sql.inserts/10seconds/<timestamp>/(.*)`, while low-resolution keys will have the form `/ts/cr.node.sql.inserts/1hour/<timestamp>/(.*)`.

In encoded keys, the `10seconds` and `1hour` segments are represented by constant string tokens. If choose the token for `1hour` to sort before the token for `10seconds`, then the newly rolled up low-resolution samples will be inserted immediately before the high-resolution samples being culled.

---

*[docs/RFCS/time_series_culling.md, line 198 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KR_IiHSXH02YhQDjcJZ-r3-198:-KTjgQF1y8XccsXPehIs:bpnlqqb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4acd5bc13a54a08259b1287a148d2db05e39887/docs/RFCS/time_series_culling.md#L198)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Those are essentially folded up into the exact same mechanism where the culling of an unculled series results in a single culled sample at the new resolution, and the culling of a culled one results in zero samples generated, right?
</details>
I'm not sure what you mean by "culled" and "unculled" series; do you mean high and low resolution series?

---

*[docs/RFCS/time_series_culling.md, line 219 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KR_IiHSXH02YhQDjcJZ-r3-219:-KTjeOVU2tu7Qr-LPtS3:b1e4kvg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4acd5bc13a54a08259b1287a148d2db05e39887/docs/RFCS/time_series_culling.md#L219)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> This isn't really correct, though I don't think we need to do anything but document that incorrectness and move on.
> 
> 
> 
> Consider that two replicas of the queue groom a time series concurrently. In reality, I think that those could even be two different replicas. For example, what if a cullable tail of a time series covers ten Ranges and each lease holder picks up the range at the same time? Many fun interleavings are possible. Let's limit to concurrency two:
> 
> 
> 
> - Replica 1 and Replica 2 get processed, discover that they have the time series.
> 
> - Replica 1 issues a Scan to get the keys to be rolled up, writes the roll-up, and deletes the keys.
> 
> - Replica 2 issues a Scan to get the keys to be rolled up. Wait, there aren't any! Let me write an empty sample and delete the keys (of which there were none).
> 
> 
> 
> Now you've created and immediately deleted the roll-up.
> 
> 
> 
> It should be very rare that two Replicas of the same Range race in this way. I'm less sure about time series which span ranges, or maybe I'm operating on false assumptions on how that works in practice.
> 
> 
> 
> In either case, there's something to discuss and document here.
</details>
Replica2 will not write empty samples if it finds no data. The system only writes samples for which keys are found.  If Replica2 finds no data, that implies that another replica has already rolled up and deleted the data, so there is nothing to do.

---

*[docs/RFCS/time_series_culling.md, line 223 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KR_IiHSXH02YhQDjcJZ-r3-223:-KTjgpIDTJrUHNKygDwd:bt3i632) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4acd5bc13a54a08259b1287a148d2db05e39887/docs/RFCS/time_series_culling.md#L223)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Ah, this seems to deal with some of it. Are we requiring that the top-boundary of the key is there? Or are we simply requiring that the result set is not empty?
</details>
Result set of what?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Thanks for clarifying, @mrtracy!

---

Review status: 0 of 1 files reviewed at latest revision, 21 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/time_series_culling.md, line 160 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KR_IiHSXH02YhQDjcJZ-r3-160:-KTkLHJQMPdWOJf0GkuA:b-7aitaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4acd5bc13a54a08259b1287a148d2db05e39887/docs/RFCS/time_series_culling.md#L160)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> They are written close to each other.  Time series keys actually have the form:
> 
> `[ts prefix][series name][*resolution token*][timestamp][source]`
> 
> Meaning that high-resolution keys have the form `/ts/cr.node.sql.inserts/10seconds/<timestamp>/(.*)`, while low-resolution keys will have the form `/ts/cr.node.sql.inserts/1hour/<timestamp>/(.*)`.
> 
> In encoded keys, the `10seconds` and `1hour` segments are represented by constant string tokens. If choose the token for `1hour` to sort before the token for `10seconds`, then the newly rolled up low-resolution samples will be inserted immediately before the high-resolution samples being culled.
</details>
Thanks for clarifying.

---

*[docs/RFCS/time_series_culling.md, line 198 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KR_IiHSXH02YhQDjcJZ-r3-198:-KTkLK-iQUiladWwLRTI:b-7f1vm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4acd5bc13a54a08259b1287a148d2db05e39887/docs/RFCS/time_series_culling.md#L198)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> I'm not sure what you mean by "culled" and "unculled" series; do you mean high and low resolution series?
</details>
Yes (or rather, low and high in the order you specify)

---

*[docs/RFCS/time_series_culling.md, line 219 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KR_IiHSXH02YhQDjcJZ-r3-219:-KTkLRmxwnRiMO6L3Yuh:b-yfdtuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4acd5bc13a54a08259b1287a148d2db05e39887/docs/RFCS/time_series_culling.md#L219)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> Replica2 will not write empty samples if it finds no data. The system only writes samples for which keys are found.  If Replica2 finds no data, that implies that another replica has already rolled up and deleted the data, so there is nothing to do.
</details>
Ok. That still leaves races (the range deletion may not be single-range, but it also can't be transactional) but I'm sure these are so rare that they're nothing to worry about, and whatever result they leave would only manage to drop parts of the data.

---

*[docs/RFCS/time_series_culling.md, line 223 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-KR_IiHSXH02YhQDjcJZ-r3-223:-KTkKyEYaEJyauyf5735:b-fqktbr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b4acd5bc13a54a08259b1287a148d2db05e39887/docs/RFCS/time_series_culling.md#L223)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> Result set of what?
</details>
Of scanning the keys which are the basis of the roll-up.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343#-:-KTkLnEdGGM0aLcw2lZV:bi1rck0)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 20 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9343)*
<!-- Sent from Reviewable.io -->
